frontend/src/DiagramEditor/DiagramEditor.jsx

```diff
@@ -178,20 +234,97 @@ export class DiagramEditor extends React.Component {
     });
   };
 
-  onComponentMouseDown = (e, selectedComponentId) => {
+  onConnectionMouseDown = (e, connectionId) => {
+    this.setState({
+      selectedConnectionId: connectionId,
+    });
+  };
+
+  onComponentMouseDown = (e, componentId) => {
+    const { isConnecting } = this.state;
+    if (!isConnecting) {
+      this.setState({
+        selectedComponentId: componentId,
+      });
+    }
+
     this.setState({
       isDraggingComponent: true,
-      selectedComponentId,
       previousMouseX: e.clientX,
       previousMouseY: e.clientY,
       initialMouseX: e.clientX,
       initialMouseY: e.clientY,
     });
   };
 
+  onComponentMouseUp = (e, componentId) => {
+    const {
+      isConnecting,
+      selectedComponentId,
+      isDraggingComponent,
+    } = this.state;
+
+    this.setState({ isDraggingComponent: false });
+
+    if (isConnecting) {
+      if (selectedComponentId !== componentId) {
+        this.props.sendChange({
+          operation: "addConnection",
+          data: {
+            from: selectedComponentId,
+            to: componentId,
+            id: `connection_${RandomWords({ exactly: 3, join: "-" })}`,
+          },
+        });
+      }
+    } else {
+      this.setState({ selectedComponentId: componentId });
+    }
+
+    if (isDraggingComponent) {
+      const selectedComponent = this.getSelectedComponent();
+
+      if (selectedComponent) {
+        const wholeDeltaX = e.clientX - (this.state.initialMouseX || 0);
+        const wholeDeltaY = e.clientY - (this.state.initialMouseY || 0);
+
+        if (!wholeDeltaX && !wholeDeltaY) {
+          console.log("onComponentMouseUp: we have no mouse move delta");
+          return;
+        }
+
+        this.setState({
```

Comment:
don't you also want to set `isDraggingComponent` to false here?

Reply:
I can't find what bit of code you're refering to here...

Reply:
ah, nvm, you're already doing it on line 267

Reply:
ah oops, now it's line 249